[Identity] Success logging (Both in Interactive and in DeviceCode)

sadasant opened the pull request:
This is a very small issue addressing #11672.

When I made #11672 I conflated two different issues related to success logging of Identity credentials (at least definitely in my mind I did conflate these two). Since the fix to these issues is very small, in number of lines changed, I thought it would be ok to make a single pull request to address them both.

In this pull request, I have:

- Changed the success message of InteractiveBrowserCredential to `Authentication Complete. You can close the browser and return to the application.`.
- Changed the success logging message of `DeviceCodeCredential` to be similar to: `DeviceCodeCredential succeeded. Scopes: {1} ExpiresOn: {3}`.
  - When I did this, I decided to change the other logging messages to make them consistent.

**Important:**
I decided to re-use the existing logging tools, so the messages are more like `ClientCertificateCredential => getToke() => SUCCESS. Scopes: {1}`

If approved,
Fixes #11672

Feedback appreciated!

sadasant commented:
Leaving this as a draft until Vinay has time to see it and give feedback!

g2vinay commented on sdk/identity/identity/src/credentials/deviceCodeCredential.ts:
```diff
@@ -127,10 +127,12 @@ export class DeviceCodeCredential implements TokenCredential {
     try {
-      return this.acquireTokenByDeviceCode(deviceCodeRequest);
+      const token = await this.acquireTokenByDeviceCode(deviceCodeRequest);
+      logger.info(`DeviceCodeCredential succeeded. Scopes: ${scopeArray.join(", ")}. Expires on timestamp: ${token?.expiresOnTimestamp}`);
```
succeeded in retrieving the access token. Scopes ? ?

sadasant replied:
I like the second one! Thank you

sadasant replied:
Did the change! Please review again.

sadasant commented:
Should we have a changelog entry for this change?

jonathandturner commented:
@sadasant - I don't think this needs its own changelog entry, though if you think so please feel free to add.

jonathandturner commented on sdk/identity/identity/src/credentials/interactiveBrowserCredential.ts:
```diff
@@ -186,7 +186,7 @@ export class InteractiveBrowserCredential implements TokenCredential {
         try {
           const authResponse = await this.pca.acquireTokenByCode(tokenRequest);
-          res.sendStatus(200);
+          res.status(200).send("Authentication Complete. You can close the browser and return to the application.");
```
Can we make a PR for just this piece? I've updated some of the success messages in https://github.com/Azure/azure-sdk-for-js/pull/11871, can you check that against this PR to see what still needs to be updated?

sadasant replied:
#11871 looks good. The caveat is that the formatSuccess method that we have uses `SUCCESS: ____` instead of `Authentication Complete`. Which one should we use, @g2vinay ?

g2vinay replied:
I'd say that the message should be consistent everywhere. Second the logging message should offer some more detail related to the status, for e.g. Successfully retrieved the access token for {Scopes}, Authentication Failed / Failed to retrieve access token for {Scopes}. These are just examples.

Does the logger in JS add more info underneath automatically beyond the message SUCCESS: {Scopes} or ERROR: {Scopes}

@schaabs

sadasant replied:
No, the logger doesn't do any magic underneath!

g2vinay commented on sdk/identity/identity/src/credentials/deviceCodeCredential.ts:
```diff
@@ -123,12 +126,22 @@ export class DeviceCodeCredential implements TokenCredential {
+          const token = await this.acquireTokenByDeviceCode(deviceCodeRequest, scopeArray);
+          logger.info(
+            formatSuccess(
+              `DeviceCodeCredential successfully retrieved the access token. Scopes: ${scopeArray.join(
```
Are other credentials, sending out an equally detailed message for success and failure logs ?

sadasant replied:
This is something I want to investigate! I can do that in this PR or in a separate PR.

sadasant commented:
@jonathandturner @g2vinay Since we already had a logging formatting tool for success and errors, I changed this PR to allow scopes in those formatting functions, and fixed the existing uses of those functions to now also send the scopes through. This will give us a more standard formatting everywhere.

Feedback appreciated!

jonathandturner commented on sdk/identity/identity/src/util/logging.ts:
```diff
@@ -47,15 +47,23 @@ export function logEnvVars(credentialName: string, supportedEnvVars: string[]):
-export function formatSuccess(scope: string | string[]): string {
-  return `SUCCESS: ${Array.isArray(scope) ? scope.join(", ") : scope}`;
+export function formatSuccess(scope: string | string[], timestamp?: number): string {
+  let message = `SUCCESS! Scopes: ${Array.isArray(scope) ? scope.join(", ") : scope}.`;
```
Please don't yell at the user. `SUCCESS:` should be okay

jonathandturner commented on sdk/identity/identity/test/internal/logger.spec.ts:
```diff
@@ -30,7 +30,7 @@ describe("Identity logging utilities", function() {
       const logger = credentialLoggerInstance("title", undefined, fakeLogger as any);
       logger.info(formatSuccess("scope"));
-      assert.equal(allParams[0].join(" "), "title => SUCCESS: scope");
+      assert.equal(allParams[0].join(" "), "title => SUCCESS! Scopes: scope.");
```
Same here, please don't change the message to a yelling message

jonathandturner commented on sdk/identity/identity/test/internal/logger.spec.ts:
```diff
@@ -40,7 +40,7 @@ describe("Identity logging utilities", function() {
       const logger = credentialLoggerInstance("title", undefined, fakeLogger as any);
       logger.info(formatSuccess(["scope 1", "scope 2"]));
-      assert.equal(allParams[0].join(" "), "title => SUCCESS: scope 1, scope 2");
+      assert.equal(allParams[0].join(" "), "title => SUCCESS! Scopes: scope 1, scope 2.");
```
ditto (I'll stop repeating this, but please fix them back)

jonathandturner commented on sdk/identity/identity/test/internal/logger.spec.ts:
```diff
@@ -110,7 +110,7 @@ describe("Identity logging utilities", function() {
     assert.equal(
       allInfoParams[1].join(" "),
-      "FakeCredential => getToken() => SUCCESS: Scope 1, Scope 2"
+      "FakeCredential => getToken() => SUCCESS! Scopes: Scope 1, Scope 2."
```
one more
